internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	targetName string
+	remoteURL  string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+	// Keep track of the number of buffered entries, to avoid having to iterate
+	// the entries map to get the total number.
+	numEntries int
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestEntries {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	err, emptyBuffer := c.handleServerResponse(resp)
```

Comment:
It's a bit weird that the err is the first rather than the second value. However, instead of switching them, why don't we just make `handleServerResponse` call `emptyBuffer` (and do the logging) directly, rather than having to return true/false and then act on it here?

That would also mean we could include the 4xx status code in the log message.